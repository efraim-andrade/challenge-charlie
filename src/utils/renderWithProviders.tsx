import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { theme } from '~/styles/theme/index';

export const renderWithProviders = (children: JSX.Element) => {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
};

export default renderWithProviders;
