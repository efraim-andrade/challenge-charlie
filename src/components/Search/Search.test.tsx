import React from 'react';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CompassIcon from '~/assets/compass.svg';
import renderWithProviders from '~/utils/renderWithProviders';

import { Search, SearchProps } from '.';

const defaultValues: SearchProps = {
  isLoading: false,
  withError: false,
  onSearch: jest.fn,
  cleanSearch: jest.fn,
};

describe('components - <Search />', () => {
  it('should render correctly', () => {
    const { container } = renderWithProviders(<Search {...defaultValues} />);

    expect(container).toMatchSnapshot();
  });

  it('should render with an icon', () => {
    renderWithProviders(
      <Search {...defaultValues} icon={{ svg: CompassIcon, alt: 'Ícone de compasso', position: 0 }} />,
    );

    expect(screen.getByLabelText('Ícone de compasso')).toBeInTheDocument();
  });

  it('should handle with change event', async () => {
    const user = userEvent.setup();

    renderWithProviders(<Search {...defaultValues} onChange={jest.fn()} />);

    const input = screen.getByRole('textbox');

    user.type(input, 'John Doe');

    jest.runAllTimers();

    await waitFor(() => {
      expect(input).toHaveValue('John Doe');
    });
  });

  it('should handle focus on input', () => {
    const user = userEvent.setup();
    const inputRef = React.createRef<HTMLInputElement>();

    renderWithProviders(<Search {...defaultValues} ref={inputRef} />);

    const input = screen.getByRole('textbox');

    if (input?.parentElement !== null) {
      user.click(input?.parentElement);
    }

    if (inputRef.current) {
      inputRef.current.focus();
    }

    expect(input).toHaveFocus();
  });
});
