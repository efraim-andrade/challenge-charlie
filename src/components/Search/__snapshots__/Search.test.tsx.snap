// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - <Search /> should render correctly 1`] = `
.c0 {
  position: relative;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.5rem;
  grid-template-columns: auto 1fr;
  width: 100%;
  padding: 1rem 0.5rem;
  opacity: 0.9;
  background: #E1E1E6;
}

.c0:has(input:focus) {
  opacity: 1;
}

.c0 img {
  -webkit-transition: 0.3s;
  transition: 0.3s;
  -webkit-transform: rotate(deg);
  -ms-transform: rotate(deg);
  transform: rotate(deg);
}

.c1 {
  border: 0;
  width: 100%;
  border: 1px solid transparent;
  border-radius: 0.5rem;
  padding: 0.5rem 5.5rem 0.5rem 0.5rem;
  font-weight: 500;
  -webkit-filter: contrast(1.2);
  filter: contrast(1.2);
  color: #7A7673;
  font-size: 2rem;
  background: #E1E1E6;
}

.c1:focus {
  outline: 0;
  border-color: #0057B7;
}

.c1::-webkit-input-placeholder {
  color: #8F8F8F;
}

.c1::-moz-placeholder {
  color: #8F8F8F;
}

.c1:-ms-input-placeholder {
  color: #8F8F8F;
}

.c1::placeholder {
  color: #8F8F8F;
}

.c2 {
  position: absolute;
  z-index: 1;
  right: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: none;
  width: 82px;
  height: 55px;
  opacity: 0.85;
  padding: 0 1rem;
  border-radius: 0 0.5rem 0.5rem 0;
  cursor: pointer;
  color: #FFF;
  font-size: 1rem;
  background: #0057B7;
  background: #8F8F8F;
}

.c2 svg {
  -webkit-animation: bcVRha 2s linear infinite;
  animation: bcVRha 2s linear infinite;
}

@media (max-width:600px) {
  .c0 {
    gap: 0.5rem;
  }

  .c0 img {
    width: 40px;
    height: 40px;
  }
}

@media (max-width:600px) {
  .c1 {
    padding: 0.5rem 4rem 0.5rem 0.5rem;
    font-size: 1.5rem;
  }
}

@media (max-width:600px) {
  .c2 {
    width: 60px;
    height: 46px;
    padding: 0 0.5rem;
  }
}

<div>
  <form
    class="c0"
  >
    <input
      class="c1"
      value=""
    />
    <button
      class="c2"
      disabled=""
      title="Buscar cidade"
      type="submit"
    >
      Buscar
    </button>
  </form>
</div>
`;
